File: riveredge-backend/src/core/schemas/realtime.py
"""实时通道 API Schema。"""

from pydantic import BaseModel, Field


class RealtimeConfigResponse(BaseModel):
    enabled: bool = Field(description="是否启用实时推送")
    backend: str = Field(default="noop", description="noop | socketio | centrifugo")
    ws_url: str = Field(
        default="",
        description="Centrifugo 专用 WS 地址；socketio 时留空，走同源",
    )
    socket_path: str = Field(
        default="/socket.io",
        description="python-socketio 路径（backend=socketio 时使用）",
    )


class RealtimeTokenResponse(BaseModel):
    token: str = Field(description="Centrifugo 连接 JWT（socketio 不需要此接口）")
    channel: str = Field(description="用户私有频道/房间名")

File: riveredge-frontend/src/services/realtime.ts
/**
 * 实时推送（REALTIME_BACKEND=socketio | centrifugo | noop）
 */

import { apiRequest } from './api';

export interface RealtimeConfig {
  enabled: boolean;
  backend: 'noop' | 'socketio' | 'centrifugo' | string;
  ws_url: string;
  socket_path: string;
}

export interface RealtimeToken {
  token: string;
  channel: string;
}

export interface RealtimeEnvelope {
  event?: string;
  payload?: Record<string, unknown>;
  ts?: string;
}

export async function getRealtimeConfig(): Promise<RealtimeConfig> {
  return apiRequest<RealtimeConfig>('/core/realtime/config', { method: 'GET' });
}

export async function getRealtimeToken(): Promise<RealtimeToken> {
  return apiRequest<RealtimeToken>('/core/realtime/token', { method: 'GET' });
}

export function resolveRealtimeWsUrl(config: RealtimeConfig): string {
  const explicit = (config.ws_url || '').trim();
  if (explicit) {
    return explicit.replace(/\/$/, '');
  }
  if (typeof window === 'undefined') {
    return '';
  }
  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
  return `${protocol}//${window.location.host}/connection/websocket`;
}

export function resolveSocketIoOrigin(): string {
  if (typeof window === 'undefined') {
    return '';
  }
  return window.location.origin;
}
